### PR TITLE
Ensure the player is on the cause stack in tryHarvestBlock().

### PR DESCRIPTION
**Pull request by JBYoshi**
This is primarily intended for the mod Ore Excavator, which calls tryHarvestBlock() from a post-tick listener. See SpongePowered/SpongeForge#2789 and SpongePowered/SpongeForge#2832.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java [hunk 1]
```diff
@@ -523,4 +527,29 @@ public EnumActionResult processRightClick(EntityPlayer player, net.minecraft.wor
...
+    private final Deque<CauseStackManager.StackFrame> tryHarvestBlockStackFrames = new LinkedList<>();
+
+    /**
+     * Some mods (e.g. OreExcavator) will call tryHarvestBlock directly, and
+     * the cause stack may not have the player on it. This should fix that.
+     * See SpongeForge#2789.
+     */
+    @Inject(method = "tryHarvestBlock", at = @At("HEAD"))
+    private void onTryHarvestBlockHead(BlockPos pos, CallbackInfoReturnable<Boolean> cir) {
+        if (!Sponge.getCauseStackManager().getCurrentCause().contains(this.player)) {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java]: Considering this code came about due to tick logic, I would rather just throw it on the stack and pop it. Worst case, we have a dupe object. Best case we don't. Either case, we save a lookup.

Comment (JBYoshi) [reply src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java]: Done.

File: src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java [hunk 2]
```diff
@@ -523,4 +527,23 @@ public EnumActionResult processRightClick(EntityPlayer player, net.minecraft.wor
         }
     }
 
+    private final Deque<CauseStackManager.StackFrame> tryHarvestBlockStackFrames = new LinkedList<>();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java]: Move this field to the top with the other mixin-added fields please

File: src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java [hunk 3]
```diff
@@ -523,4 +527,23 @@ public EnumActionResult processRightClick(EntityPlayer player, net.minecraft.wor
...
+
+    /**
+     * Some mods (e.g. OreExcavator) will call tryHarvestBlock directly, and
+     * the cause stack may not have the player on it. This should fix that.
+     * See SpongeForge#2789.
+     */
+    @Inject(method = "tryHarvestBlock", at = @At("HEAD"))
+    private void onTryHarvestBlockHead(BlockPos pos, CallbackInfoReturnable<Boolean> cir) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java]: with the ongoing cleanup/splitting of mixins, can you name this `impl$addPlayerToCauseStackForHarvesting`?

File: src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java [hunk 4]
```diff
@@ -523,4 +527,23 @@ public EnumActionResult processRightClick(EntityPlayer player, net.minecraft.wor
...
+        this.tryHarvestBlockStackFrames.push(Sponge.getCauseStackManager().pushCauseFrame());
+        Sponge.getCauseStackManager().pushCause(this.player);
+    }
+
+    @Inject(method = "tryHarvestBlock", at = @At("RETURN"))
+    private void onTryHarvestBlockReturn(BlockPos pos, CallbackInfoReturnable<Boolean> cir) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerInteractionManager.java]: And this `impl$PopPlayerFromStackForHarvesting`

**Review by gabizou — CHANGES_REQUESTED** [review]
I'm debating on the whole whether we'd want to enter/exit phase states so that we'd avoid potentially forcing a stackframe creation (since stackframes are lazily initialized based on the phase states). PhaseState entry/exit is relatively cheap, and adding a new class here as generalized as it is, will allow us to potentially refine the entry to support whether using the `CapturePlayer` object, or whether we'd just write it as a new state providing the frame modifier.

Alternatively, can check on the phase state to default to adding the player as a modifier, or adding the player to the phase context as a captured player during harvesting sort of deal.

**Comment by Zidane** [comment]
Can you update this PR @JBYoshi ?

**Comment by JBYoshi** [comment]
@Zidane Updated.

**Comment by gabizou** [comment]
Merged in to SF to use a mixin there so it's not always occurring in SpongeVanilla.

**Pull request closed without merging**